src/hooks/mocha.ts

```diff
@@ -0,0 +1,66 @@
+import { simple as walk } from "acorn-walk";
+import type { ESTree } from "meriyah";
+import { expressionFor } from ".";
+import { call_, identifier } from "../generate";
+import { EventEmitter } from "node:stream";
+import Recording from "../Recording";
+import { recording, start } from "../recorder";
+
+export function shouldInstrument(url: URL): boolean {
+  return url.pathname.endsWith("/mocha/lib/runner.js");
+}
+
+export function transform(program: ESTree.Program): ESTree.Program {
+  return patchRuntime(program);
+}
+
+const EVENT_TEST_BEGIN = 'test';
+const EVENT_TEST_PASS = 'pass';
+const EVENT_TEST_FAIL = 'fail';
+const EVENT_TEST_END = 'test end';
+
+interface Test {
+  title: string
+  fullTitle: () => string
+}
+
+function registerEventListeners(runner: EventEmitter) {
+  runner.on(EVENT_TEST_BEGIN, function (test: Test) {
+    const rec = new Recording("tests", "mocha", test.fullTitle());
+    start(rec);
+  });
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  runner.on(EVENT_TEST_PASS, function (test: Test) {
+    recording.metadata.test_status = "succeeded";
+    recording.finish();
+  });
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  runner.on(EVENT_TEST_FAIL, function (test: Test) {
+    recording.metadata.test_status = "failed";
+    recording.finish();
+  });
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  runner.on(EVENT_TEST_END, function (test: Test) {
+    ;
+  });
+}
+
+export function patchRuntime(program: ESTree.Program): ESTree.Program {
+  recording.abandon();
```

Comment:
I feel this `abandon` should happen only after we successfully patch mocha (so, at the end of the `if` in the MethodDefinition walker). A message to the user about mocha being detected is also needed.

Reply:
You've missed the user message. I'll add that.